server/sonar-server/src/main/java/org/sonar/server/es/queue/RecoveryIndexer.java

```diff
@@ -0,0 +1,151 @@
+  private static final long DEFAULT_DELAY_IN_SECONDS = 5L * 60;
+  private static final long DEFAULT_MIN_AGE_IN_SECONDS = 5L * 60;
+  private static final double FAILURE_CIRCUIT_BREAKER_IN_PERCENT = 0.3;
+  private static final int LOOP_LIMIT = 10_000;
+
+  private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1,
+    new ThreadFactoryBuilder()
+      .setPriority(Thread.MIN_PRIORITY)
+      .setNameFormat("RecoveryIndexer-%d")
+      .build());
+  private final System2 system2;
+  private final Settings settings;
+  private final DbClient dbClient;
+  private final UserIndexer userIndexer;
+
+  // initialized at startup
+  private long minAgeInSeconds;
+
+  public RecoveryIndexer(System2 system2, Settings settings, DbClient dbClient, UserIndexer userIndexer) {
+    this.system2 = system2;
+    this.settings = settings;
+    this.dbClient = dbClient;
+    this.userIndexer = userIndexer;
+  }
+
+  @Override
+  public void start() {
+    long delayInSeconds = getSetting(PROPERTY_DELAY, DEFAULT_DELAY_IN_SECONDS);
+    minAgeInSeconds = getSetting(PROPERTY_MIN_AGE, DEFAULT_MIN_AGE_IN_SECONDS);
+
+    // in the cluster mode, avoid (but not prevent!) simultaneous executions of recovery
+    // indexers so that a document is not handled multiple times.
+    int initialDelay = RandomUtils.nextInt((int) (delayInSeconds / 2));
+
+    executorService.scheduleAtFixedRate(
+      this::index,
+      initialDelay,
+      delayInSeconds,
+      TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void stop() {
+    executorService.shutdown();
+  }
+
+  protected void index() {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Profiler profiler = Profiler.create(LOG).start();
+      long beforeDate = system2.now() - minAgeInSeconds * 1_000;
+      long total = 0L;
+      long totalSuccess = 0L;
+
+      Collection<EsQueueDto> items = dbClient.esQueueDao().selectForRecovery(dbSession, beforeDate, LOOP_LIMIT);
+      while (!items.isEmpty()) {
+        total += items.size();
+        long loopSuccess = 0L;
+
+        ListMultimap<EsQueueDto.Type, EsQueueDto> itemsByType = groupItemsByType(items);
+        for (Map.Entry<EsQueueDto.Type, Collection<EsQueueDto>> entry : itemsByType.asMap().entrySet()) {
+          loopSuccess += doIndex(dbSession, entry.getKey(), entry.getValue());
+        }
+
+        totalSuccess += loopSuccess;
+        if (loopSuccess / items.size() < FAILURE_CIRCUIT_BREAKER_IN_PERCENT) {
+          LOG.error("Elasticsearch recovery - too many failures [{}/{} documents], waiting for next run", items.size() - loopSuccess, items.size());
+          break;
+        }
+        items = dbClient.esQueueDao().selectForRecovery(dbSession, beforeDate, LOOP_LIMIT);
+      }
+      if (total > 0L) {
+        profiler.stopInfo(format("Elasticsearch recovery - %d documents processed [%d failures]", total, total - totalSuccess));
+      }
+    } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)